File: src/game/inner/piece.rs
use super::{field::Cell, grid::Grid, input::Turn};

#[derive(Clone, Copy, Debug)]
pub struct Piece {
    ty: PieceType,
    rotations: u8,
}

#[derive(Clone, Copy, Debug)]
enum PieceType {
    Square,
    Stick,
    L,
    ReverseL,
    T,
    S,
    ReverseS,
}

const fn from_bool(b: bool) -> Cell {
    if b {
        Cell::Falling
    } else {
        Cell::Empty
    }
}

const fn from_bin(bin: u8) -> [Cell; 4] {
    [
        from_bool(bin & 0b_1000 != 0),
        from_bool(bin & 0b_0100 != 0),
        from_bool(bin & 0b_0010 != 0),
        from_bool(bin & 0b_0001 != 0),
    ]
}

fn blueprint(bs: [u8; 4]) -> Grid<Cell, 4, 4> {
    [
        from_bin(bs[0]),
        from_bin(bs[1]),
        from_bin(bs[2]),
        from_bin(bs[3]),
    ]
    .into()
}

impl Piece {
    pub fn new() -> Self {
        Self {
            ty: PieceType::Square,
            rotations: 0,
        }
    }

    pub fn next(self) -> Self {
        let mut result = Self::new();
        result.ty = {
            match (self.ty as usize).wrapping_add(1) % 7 {
                0 => PieceType::Square,
                1 => PieceType::Stick,
                2 => PieceType::L,
                3 => PieceType::ReverseL,
                4 => PieceType::T,
                5 => PieceType::S,
                6 => PieceType::ReverseS,
                _ => unreachable!(),
            }
        };
        result
    }

    pub fn turn(&mut self, turn: Turn) {
        self.rotations = match turn {
            Turn::Left => self.rotations.wrapping_sub(1) % 4,
            Turn::Right => self.rotations.wrapping_add(1) % 4,
        };
    }

    pub fn dims(self) -> [usize; 2] {
        let blueprint = self.blueprint();

        let [y0, x0] = [0, 0];
        let [y1, x1] = [blueprint.height(), blueprint.width()];

        let mut piece_width = blueprint.width();
        let mut piece_height = blueprint.height();
        for y in (y0..y1).rev() {
            if (x0..x1).any(|x| blueprint[[y, x]] == Cell::Falling) {
                piece_height = y + 1;
                break
            }
        }
        for x in (x0..x1).rev() {
            if (y0..y1).any(|y| blueprint[[y, x]] == Cell::Falling) {
                piece_width = x + 1;
                break
            }
        }
        [piece_height, piece_width]
    }

    pub fn blueprint(self) -> Grid<Cell, 4, 4> {
        let mut blueprint = match self.ty {
            PieceType::Square => blueprint([
                0b_1100,
                0b_1100,
                0b_0000,
                0b_0000,
            ]),
            PieceType::Stick => blueprint([
                0b_1111,
                0b_0000,
                0b_0000,
                0b_0000,
            ]),
            PieceType::L => blueprint([
                0b_1000,
                0b_1000,
                0b_1100,
                0b_0000,
            ]),
            PieceType::ReverseL => blueprint([
                0b_0100,
                0b_0100,
                0b_1100,
                0b_0000,
            ]),
            PieceType::T => blueprint([
                0b_0100,
                0b_1110,
                0b_0000,
                0b_0000,
            ]),
            PieceType::S => blueprint([
                0b_0110,
                0b_1100,
                0b_0000,
                0b_0000,
            ]),
            PieceType::ReverseS => blueprint([
                0b_1100,
                0b_0110,
                0b_0000,
                0b_0000,
            ]),
        };

        for _ in 0..self.rotations {
            blueprint = blueprint.rotate(false);
        }

        {
            let [width, height] = [blueprint.width(), blueprint.height()];

            let mut y0 = 0;
            for y in 0..height {
                if (0..width).any(|x| blueprint[[y, x]] == Cell::Falling) {
                    y0 = y;
                    break
                }
            }
            let mut x0 = 0;
            for x in 0..width {
                if (0..height).any(|y| blueprint[[y, x]] == Cell::Falling) {
                    x0 = x;
                    break
                }
            }

            let mut result = blueprint;
            result.iter_mut().for_each(|c| *c = Cell::Empty);
            for (dst_y, src_y) in (y0..height).enumerate() {
                for (dst_x, src_x) in (x0..width).enumerate() {
                    result[[dst_y, dst_x]] = blueprint[[src_y, src_x]];
                }
            }
            blueprint = result;
        }

        blueprint
    }
}

File: haskell/app/Game/Piece.hs
module Game.Piece
  ( Piece
  , mkPiece
  , PieceType (Square, Stick, L, ReverseL, T, S, ReverseS)
  , dims
  , rowsCols
  , blueprint
  , turn
  , next
  ) where

import GHC.Arr
import Data.Bits ((.&.))
import Data.List
import Data.List.Split
import Data.Maybe
import Data.Word

import Game.Field hiding (dims)
import Game.Field qualified as Field (dims)
import Game.Input

data Piece = Piece
  { ty :: PieceType
  , rotations :: Int
  } deriving Show

data PieceType
  = Square
  | Stick
  | L
  | ReverseL
  | T
  | S
  | ReverseS
  deriving (Show, Enum, Bounded)

mkPiece :: Piece
mkPiece = Piece
  { ty        = Square
  , rotations = 0
  }

next :: Piece -> Piece
next p = p
  { ty        = toEnum (succ (fromEnum (ty p)) `mod` (fromEnum (maxBound :: PieceType) + 1))
  , rotations = 0
  }

-- TODO: mod or %?
turn :: Turn -> Piece -> Piece
turn TurnLeft  p = p { rotations = pred (rotations p) `mod` 4 }
turn TurnRight p = p { rotations = succ (rotations p) `mod` 4 }

dims :: Piece -> (Int, Int)
dims p =
  let
    b      = blueprint p
    (h, w) = Field.dims b
    (rows, cols) = rowsCols b
  in 
    ( maybe h (h-) . findIndex (elem Falling) . reverse $ rows
    , maybe w (w-) . findIndex (elem Falling) . reverse $ cols
    )

rowsCols :: Array Pos Cell -> ([[Cell]], [[Cell]])
rowsCols a =
  ( chunksOf w . map snd .               assocs $ a
  , chunksOf h . map snd . sortOn cols . assocs $ a
  )
  where
    (h, w) = Field.dims a
    cols (Pos (_, x), _) = x

blueprint :: Piece -> Array Pos Cell
blueprint p =
  let
    rotated = iterate (rotate False) bp !! rotations p
    (srcY, srcX) =
      let (rows, cols) = rowsCols rotated
      in
        ( fromMaybe 0 $ findIndex (elem Falling) rows
        , fromMaybe 0 $ findIndex (elem Falling) cols
        )
    getOrEmpty i a = if inRange (bounds a) i then a ! i else Empty
  in
    array (bounds rotated)
      [(Pos (y, x), getOrEmpty (Pos (y + srcY, x + srcX)) rotated) | Pos (y, x) <- range (bounds rotated)]
  where 
    bp = mkBlueprint $ case ty p of
      Square   -> [ 0b_1100
                  , 0b_1100
                  , 0b_0000
                  , 0b_0000
                  ]
      Stick    -> [ 0b_1111
                  , 0b_0000
                  , 0b_0000
                  , 0b_0000
                  ]
      L        -> [ 0b_1000
                  , 0b_1000
                  , 0b_1100
                  , 0b_0000
                  ]
      ReverseL -> [ 0b_0100
                  , 0b_0100
                  , 0b_1100
                  , 0b_0000
                  ]
      T        -> [ 0b_0100
                  , 0b_1110
                  , 0b_0000
                  , 0b_0000
                  ]
      S        -> [ 0b_0110
                  , 0b_1100
                  , 0b_0000
                  , 0b_0000
                  ]
      ReverseS -> [ 0b_1100
                  , 0b_0110
                  , 0b_0000
                  , 0b_0000
                  ]

mkBlueprint :: [Word8] -> Array Pos Cell
mkBlueprint bs = listArray (Pos (0, 0), Pos (3, 3)) $ concatMap fromBin bs

fromBin :: Word8 -> [Cell]
fromBin bin = map (fromBool . (/= 0) . (.&. bin)) [ 0b_1000
                                                  , 0b_0100
                                                  , 0b_0010
                                                  , 0b_0001
                                                  ]

fromBool :: Bool -> Cell
fromBool True  = Falling
fromBool False = Empty
    